src/com/sun/jna/ParameterTypeMapper.java

```diff
@@ -0,0 +1,57 @@
+/* Copyright (c) 2020 Torbjörn Svensson, All Rights Reserved
+ *
+ * The contents of this file is dual-licensed under 2
+ * alternative Open Source/Free licenses: LGPL 2.1 or later and
+ * Apache License 2.0. (starting with JNA version 4.0.0).
+ *
+ * You can freely decide which license you want to apply to
+ * the project.
+ *
+ * You may obtain a copy of the LGPL License at:
+ *
+ * http://www.gnu.org/licenses/licenses.html
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ *
+ * You may obtain a copy of the Apache License at:
+ *
+ * http://www.apache.org/licenses/
+ *
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+/**
+ * Map a function parameter to a new type
+ *
+ * @author Torbj&ouml;rn Svensson, azoff[at]svenskalinuxforeninen.se
+ */
+@Documented
+@Retention(RUNTIME)
+@Target(METHOD)
+public @interface ParameterTypeMapper {
```

Comment:
The idea is good, but not with the Java6 requirement. We can discuss whether that should be dropped, but fixing problems by introducing a new source of bugs is IMHO a no-go.

If this is done, it also gets more complex, as I can imagine situations where parameter mapping needs pre-call and post-call methods.

Reply:
This is Java6 compliant, at least the ant rules passes with it.
I think you are referring to annotations on a parameter that was introduced in Java8, but this annotation interface works around the limitations of Java6 by applying the annotation on the method instead.

Do you have some other solution to provide a simple java.lang.String in Java while the data provided in the native code is either a char* (ANSI) or a wchar_t* (UNICODE)? Keep in mind that this should not be applied where the normal rules does not apply, like where there is another parameter that tells the native layer what encoding that is applied to the binary data provided.

Reply:
Yes it is Java6 compliant, but people will fail to use this correctly and thus will introduce subtle bugs. The solution could be simple:

`Advapi32 INSTANCE = Native.load("Advapi32", Advapi32.class, W32APIOptions.UNICODE);`

This brings us back into save territory. It is not established, that there is even a problem out there. Setups with `w32.ascii` are already broken and failed to notice, so maybe there is no problem? But this again enters "what is the right solution" territory, that should be taken to the mailinglist.